### ISSUE TITLE
Compile model: generate code for query filters when possible

### ISSUE THREAD

**Issue by AndriySvyryd** [dotnet/efcore#24897]
Avoid the need for post-processing for relational providers

**Comment by stevendarby** [comment]
Would be interested to know if you have a rough idea of what sort of query filters you might be able to generate code for? Just to get any idea if our model with query filters is likely to be supported in future. The filters are like `x.TenantId == this.TenantId && x.SoftDeleted == false`

**Comment by ajcvickers** [comment]
Note from triage: consider encapsulating filter in a type as a way of not needing the expression tree to be contained in the model.

**Comment by ilmax** [comment]
It would be awesome to get more compiled models features in EF 8, this one is in particular since query filters are quite useful, so I tend to use them a lot

**Comment by zejji** [comment]
Is there a workaround for this in the interim, i.e. is there a way to use compiled models together with global query filters even if code for query filters is not currently generated? The original post refers to avoiding the need for "post-processing", but I'm not sure what this is referring to.

**Comment by ajcvickers** [comment]
@zejji I don't believe there is any workaround for this.

**Comment by truonghoangsw** [comment]
We are add globle filter. In Function CreateAnnotations of file EntityType
` public static void CreateAnnotations(RuntimeEntityType runtimeEntityType)
        {
            runtimeEntityType.AddAnnotation("Relational:FunctionName", null);
            runtimeEntityType.AddAnnotation("Relational:Schema", null);
            runtimeEntityType.AddAnnotation("Relational:SqlQuery", null);
            runtimeEntityType.AddAnnotation("Relational:TableName", "Application");
            runtimeEntityType.AddAnnotation("Relational:ViewName", null);
            runtimeEntityType.AddAnnotation("Relational:ViewSchema", null);
			Expression<Func<Application, bool>> expression = Application => !Application.IsDeleted;
			runtimeEntityType.AddAnnotation(CoreAnnotationNames.QueryFilter, expression);
			Customize(runtimeEntityType);
        }` can i use it like this? @ajcvickers

**Comment by tidan-16** [comment]
Would be very nice for ef core 9! Most of the time we would want to use it for multi tenancy and soft deleting

**Comment by stevendarby** [comment]
> Would be very nice for ef core 9! Most of the time we would want to use it for multi tenancy and soft deleting

Unfortunately there is no clear scope defined for what "when possible" means, so these scenarios might not be supported.

**Comment by tidan-16** [comment]
truonghoangsw's comment seems to work no?

**Comment by truonghoangsw** [comment]
> truonghoangsw's comment seems to work no?

Yes,it work

**Comment by klemmchr** [comment]
I suppose that this issue is about the limitation of using global query filters in combination with compiled models. The [original announcement](https://devblogs.microsoft.com/dotnet/announcing-entity-framework-core-6-0-preview-5-compiled-models/) states that it is not possible to use compiled models with global query filters. Is this still the case? I'm unable to find any first party documentation about compiled models.

I would like to use compiled models with global query filters. Currently we are using them to filter out soft deleted entities. Some comments here mention that this is possible. Is this the case?

**Comment by AndriySvyryd** [comment]
@klemmchr Yes, as a workaround you can remove the global filter, generate a compiled model and then manually edit it to add the filter as seen in https://github.com/dotnet/efcore/issues/24897#issuecomment-1670643084

**Comment by klemmchr** [comment]
> @klemmchr Yes, as a workaround you can remove the global filter, generate a compiled model and then manually edit it to add the filter as seen in [#24897 (comment)](https://github.com/dotnet/efcore/issues/24897#issuecomment-1670643084)

Would this change per persisted between model updates and recompilations or do I need to add this manually every time after compiling?

**Comment by AndriySvyryd** [comment]
> Would this change be persisted between model updates and recompilations or do I need to add this manually every time after compiling?

You'd need to do it every time you regenerate the model.

**Comment by erdalsivri** [comment]
You can do it in the Customize partial method in a separate file. Compiled model EntityType classes are all partial

**Comment by rock-it-rob** [comment]
Is the workaound posted here by @truonghoangsw still valid? Do we need to worry about this warning?

<img width="956" height="280" alt="Image" src="https://github.com/user-attachments/assets/cb844e3a-7de3-4499-9238-5a3d7cc7d723" />

**Comment by matthetherington** [comment]
We're using named query filters and the workaround above (https://github.com/dotnet/efcore/issues/24897#issuecomment-1670643084) didn't work without some modification. I now have this:

```csharp
namespace MyNamespace.CompiledModels;

[SuppressMessage("Usage", "EF1001:Internal EF Core API usage.")]
public partial class MyClassEntityType
{
    static partial void Customize(RuntimeEntityType runtimeEntityType)
    {
        Expression<Func<MyClass, bool>> expression = q => q.DeletedAt == null;
        QueryFilterCollection queryFilterCollection = new();
        RuntimeQueryFilter queryFilter = new RuntimeQueryFilter(
            QueryFilterNames.FilterSoftDeleted,
            expression
        );
        queryFilterCollection.Set(queryFilter);
        runtimeEntityType.AddAnnotation(CoreAnnotationNames.QueryFilter, queryFilterCollection);
    }
}
```
